### PR TITLE
armaOS command changes and additions

### PR DESCRIPTION
**Pull request by y0014984**
- added 'echo' command
- fixed option count checks
- harmonized option count checks
- harmonized function headers
- added log file support
- added security log file (login, failed login etc.)
- changed initial filesystem objects loading from script to config
- added multipline output for 'cat'
- added writeToFile function
- added /sys/battery/capacity filesystem object for later use
- added 'crypto' command
- added 'crack' command
- added 'find'command

20221104114755_1_cropped

### CODE REVIEW

**Comment by y0014984** [comment]
Started adding a `find` command but cam to an halt, because our `AE3_Filesystem` is very weird:
``` sqf
private _filesystem = cursorObject getVaraible "AE3_Filesystem";

typeName _filesystem; // ARRAY
typeName (_filesystem select 0); // HASHMAP
typeName ((_filesystem select 0) get "bin"); // ARRAY
```
I can't use the `in` command if the type is changing and and want to go through the filesystem recursively. That folders like `bin` are Arrays could happen due to my change of folder creation. I use the `createDir` function that inserts Arrays instead of Hashmaps. I need to talk about that with @GermanHydrogen

**Comment by y0014984** [comment]
Perhaps it's always a single Hashmap in an Array like, `[createHashMap]`. And perhaps every Hashmap Item also contains a  `[createHashMap]` for subitems.

File: addons/filesystem/functions/fnc_appendToFile.sqf [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+private _current = _dir select 1;
+_current = _current select 0;
+
+private _file = _dir select 2;
+
+if(!(_file in _current)) throw (format ["'%1' not found!", _file]);
+
+private _permissionWrite = 2; // write permission needed to append to file
+[_current get _file, _user, _permissionWrite] call AE3_filesystem_fnc_hasPermission;
+
```
Comment (GermanHydrogen) [inline addons/filesystem/functions/fnc_appendToFile.sqf:38]: Probably better:
```
private _file_object = _current get _file;
_file_object  set [0, (_file_object select 0) + _appendableContent]


```

File: addons/filesystem/functions/fnc_findFilesystemObject.sqf [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+			_totalResults append _subResults;
+		};
+	} forEach _content;
+
+} catch { _missingPermissions = _missingPermissions + 1; };
+
+[_totalResults, _missingPermissions];
+
+// TODOS:
```
Comment (GermanHydrogen) [inline addons/filesystem/functions/fnc_findFilesystemObject.sqf:84]: Todos should be kept as issues

File: addons/filesystem/functions/fnc_findFilesystemObject.sqf [hunk 3]
```diff
@@ -0,0 +1,86 @@
+/**
+ * Searches for a filesystem object in a given directory. 
+ *
+ * Arguments:
+ * 0: Pointer <[STRING]>
+ * 1: Filesystem [<HASHMAP>, <STRING>]
+ * 4: User <STRING>
```
Comment (GermanHydrogen) [inline addons/filesystem/functions/fnc_findFilesystemObject.sqf:7]: ?

File: addons/filesystem/functions/fnc_writeToFile.sqf [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+_current = _current select 0;
+
+private _file = _dir select 2;
+
+if(!(_file in _current)) throw (format ["'%1' not found!", _file]);
+
+private _permissionWrite = 2; // write permission needed to write to file
+[_current get _file, _user, _permissionWrite] call AE3_filesystem_fnc_hasPermission;
+
+private _owner = (_current get _file) select 1;
```
Comment (GermanHydrogen) [inline addons/filesystem/functions/fnc_writeToFile.sqf:39]: Same as above:
```
private _file_object = _current get _file;
_file_object  set [0, _content]
```

File: addons/filesystem/functions/fnc_initFilesystem.sqf [hunk 5]
```diff
@@ -8,30 +8,58 @@
...
+private _filesystemObjects = ("inheritsFrom _x == (configFile >> 'AE3_FilesystemObject')" configClasses _config);
 
-	// Root home dir
-	["root", [createHashMap, 'root', [[true, true, true], [false, false, false]]]],
+{
+	private _ptr = [];
 
-	// Mount file for tmp filesystems
-	["mnt", [createHashMap, 'root', [[true, true, true], [true, true, false]]]],
+	try
```
Comment (GermanHydrogen) [inline addons/filesystem/functions/fnc_initFilesystem.sqf:26]: Why try here?

**Review by GermanHydrogen — CHANGES_REQUESTED** [review]
Please a push request for each (unrelated) feature you are introducing. This would
make testing and the code review much easier.

File: addons/armaos/functions/fnc_encryption_caesar.sqf [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+private _allowedAlphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"; // 26 Characters
+
+// convert all characters to upper case
+_message = toUpper _message;
+
+// filter message with allowed characters
+_message = [_message, _allowedAlphabet + " "] call BIS_fnc_filterString;
+
+// normalize key to be lower then alphabet count
+if (_key >= (count _allowedAlphabet)) then { _key = _key % (count _allowedAlphabet)}; // 26 == 0
```
[the remark this answers is not in the document]
Comment (y0014984) [reply addons/armaos/functions/fnc_encryption_caesar.sqf:29]: You are right. I will change that.

File: addons/armaos/functions/fnc_encryption_caesar.sqf [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+{
+    private _encryptionIndex = _allowedAlphabet find _x;
+
+    // ignore spaces for conversion
+    if (!(_x isEqualTo " ")) then
+    {
+        if (_mode isEqualTo "encrypt") then
+        {
+            // check if index out of bounds and encrypt
+            if ((_encryptionIndex + _key) > ((count _allowedAlphabet) -1)) then
```
[the remark this answers is not in the document]
Comment (y0014984) [reply addons/armaos/functions/fnc_encryption_caesar.sqf:47]: The key is normalized in line 29, so there are no higher key values. But I will change that like you suggested.

File: addons/armaos/functions/fnc_os_crack.sqf [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * Results:
+ * None
+ */
+
+params ["_computer", "_options"];
+
+if (count _options < 2) exitWith { [_computer, "'crack' has too few options"] call AE3_armaos_fnc_shell_stdout; };
+
+// parse options --> perhaps we should develop a general options parsing algorythm
```
[the remark this answers is not in the document]
Comment (y0014984) [reply addons/armaos/functions/fnc_os_crack.sqf:17]: done #207

File: addons/armaos/functions/fnc_os_crack.sqf [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+private _algorythm = "";
+private _mode = "";
+private _message = "";
+
+// recognized options are overwritten by "" so the resulting string is the message to process; No need for quotations
+{
+    if (_x isEqualTo "-a") then { _algorythm = _options select (_forEachIndex + 1); _options set [_forEachIndex, ""]; _options set [_forEachIndex + 1, ""]; };
+    if (_x isEqualTo "-m") then { _mode = _options select (_forEachIndex + 1); _options set [_forEachIndex, ""]; _options set [_forEachIndex + 1, ""]; };
+} forEach _options;
+
```
[the remark this answers is not in the document]
Comment (y0014984) [reply addons/armaos/functions/fnc_os_crack.sqf:27]: I will add additional checks.

File: addons/armaos/functions/fnc_os_crack.sqf [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+            _statistics set [_x, _count];
+        };
+    } forEach (_message splitString "");
+
+    private _foundChars = keys _statistics;
+
+    _foundChars sort true;
+
+    {
+        private _indexOfE = _allowedAlphabet find "E";
```
[the remark this answers is not in the document]
Comment (y0014984) [reply addons/armaos/functions/fnc_os_crack.sqf:75]: You are right. Will change that.

File: addons/armaos/functions/fnc_os_crack.sqf [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        private _keyIfThisIsAnE = _allowedAlphabet find _x;
+        if (_keyIfThisIsAnE >= _indexOfE) then
```
[the remark this answers is not in the document]
Comment (y0014984) [reply addons/armaos/functions/fnc_os_crack.sqf:77]: I think the solution is something like: `_keyIfThisIsAnE = ((_keyIfThisIsAnE + (count _allowedAlphabet)) - _indexOfE) % (count _allowedAlphabet);`

File: addons/filesystem/functions/fnc_appendToFile.sqf [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+ * --------
+ * private _pointer = cursorObject getVariable "AE3_Filepointer";
+ * private _filesystem = cursorObject getVariable "AE3_Filesystem";
+ * private _filePath = "/tmp/new/example.txt";
+ * private _user = "root"; 
+ * private _appendableContent = "\n" + "new line";
+ * [_pointer, _filesystem, _filePath, _user, _appendableContent] call AE3_filesystem_fnc_appendToFile;
+*/
+
+// Perhaps this function could be combined with the writeToFile function. append could be a BOOL optional parameter.
```
[the remark this answers is not in the document]
Comment (y0014984) [reply addons/filesystem/functions/fnc_appendToFile.sqf:24]: Will do.

File: addons/filesystem/functions/fnc_appendToFile.sqf [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+private _dir = [_pntr, _filesystem, _target, _user] call AE3_filesystem_fnc_getParentDir;
+private _current = _dir select 1;
+_current = _current select 0;
+
+private _file = _dir select 2;
+
+if(!(_file in _current)) throw (format ["'%1' not found!", _file]);
+
+private _permissionWrite = 2; // write permission needed to append to file
```
[the remark this answers is not in the document]
Comment (y0014984) [reply addons/filesystem/functions/fnc_appendToFile.sqf:36]: There is technically no need for this variable but it was easier for me to read what this value means. I will change that to a comment.

File: addons/filesystem/functions/fnc_appendToFile.sqf (hunk 1 above)
Comment (y0014984) [reply addons/filesystem/functions/fnc_appendToFile.sqf:38]: I will change that.

File: addons/filesystem/functions/fnc_findFilesystemObject.sqf (hunk 2 above)
Comment (y0014984) [reply addons/filesystem/functions/fnc_findFilesystemObject.sqf:84]: Already done in #193. I will remove the comments.

File: addons/filesystem/functions/fnc_initFilesystem.sqf (hunk 5 above)
Comment (y0014984) [reply addons/filesystem/functions/fnc_initFilesystem.sqf:26]: Perhaps a testing artefact. I will remove that.

**Review by GermanHydrogen — APPROVED** [review]

**Pull request merged**